measure_optimize.py

```diff
@@ -2,180 +2,2412 @@
+        Shape: sum(Nt) * sum(Nt)
+    """
+
+    identity = 0
+    variance = 1
+    time_correlation = 2
+    measure_correlation = 3
+    time_measure_correlation = 4
+
+    @classmethod
+    def has_value(cls, value):
+        return value in cls._value2member_map_
+
+
+class ObjectiveLib(Enum):
+    """
+    Objective function library
+
+    if A: minimize the trace of FIM
+    if D: minimize the determinant of FIM
+    """
+
+    A = 0
+    D = 1
+
+    @classmethod
+    def has_value(cls, value):
+        return value in cls._value2member_map_
+
+
+class SensitivityData:
+    """"""
+
+    def __init__(self, filename, Nt) -> None:
+        """
+        Read, convert the format, and sotre the sensitivity information (Jacobian).
+        Only process a certain format of CSV file.
+        Assume every measurement has the same number of time points: Nt
+
+        Arguments
+        ---------
+        :param filename: a string of the csv file name
+        :param Nt: number of timepoints is needed to split Q for each measurement.
+            Assume all measurements have the same number of time points of Nt.
+
+        This csv file should have the following format:
+        columns: parameters to be estimated
+        (An extra first column is added for index)
+        rows: measurement timepoints
+        data: Jacobian values
+
+        The csv file example:
+
+        Column index  |  Parameter 1 | Parameter 2 |  ...  | Parameter P
+        measurement 1 |    number    |  number     |  ...  | number
+        measurement 2 |    number    |  number     |  ...  | number
+        ...
+        measurement N |    number    |  number     |  ...  | number
+
+        Number according to measurement i, parameter j is
+        the gradient of measurement i w.r.t parameter j
+
+        Returns
+        ------
+        None
+        """
+        # store the number of time points for each measurement
+        self.Nt = Nt
+        # read Jacobian from .csv file
+        self._read_from_csv(filename)
+
+    def _read_from_csv(self, filename):
+        """Read Jacobian from csv file
+
+        Arguments
+        ---------
+        :param filename: a string of the csv file name
+            This csv file structure is explained in the class doc string.
+
+        Returns
+        ------
+        None
+        """
+        jacobian_info = pd.read_csv(filename, index_col=False)
+        # convert Jacobian matrix from Pandas DataFrame to numpy array
+        jacobian_array = np.asarray(jacobian_info)
+
+        # check if the number of rows in this array is a multiple of Nt
+        if len(jacobian_array) % self.Nt != 0:
+            raise ValueError(
+                "The number of rows in this Jacobian matrix is not a multiple of Nt."
+            )
+
+        # infer the number of parameters from the shape of jacobian
+        # the first column is the measurement name, so needs to be deleted
+        self.n_parameters = len(jacobian_array[0]) - 1
```

Comment:
Is this needed is you set index_col=True when reading the csv?

Reply:
No, this is no longer needed. I changed this part.